include/pybind11/numpy.h

```diff
@@ -98,31 +121,43 @@ class array : public buffer {
     }
 
     array(const buffer_info &info) {
+        PyObject *arr = nullptr, *descr = nullptr;
+        int ndim = 0;
+        Py_ssize_t dims[32];
         API& api = lookup_api();
-        if ((info.format.size() < 1) || (info.format.size() > 2))
-            pybind11_fail("Unsupported buffer format!");
-        int fmt = (int) info.format[0];
-        if (info.format == "Zd")      fmt = API::NPY_CDOUBLE_;
-        else if (info.format == "Zf") fmt = API::NPY_CFLOAT_;
-
-        PyObject *descr = api.PyArray_DescrFromType_(fmt);
-        if (descr == nullptr)
-            pybind11_fail("NumPy: unsupported buffer format '" + info.format + "'!");
-        object tmp(api.PyArray_NewFromDescr_(
-            api.PyArray_Type_, descr, (int) info.ndim, (Py_intptr_t *) &info.shape[0],
-            (Py_intptr_t *) &info.strides[0], info.ptr, 0, nullptr), false);
-        if (info.ptr && tmp)
-            tmp = object(api.PyArray_NewCopy_(tmp.ptr(), -1 /* any order */), false);
-        if (!tmp)
-            pybind11_fail("NumPy: unable to create array!");
-        m_ptr = tmp.release().ptr();
+
+        // Allocate zeroed memory if it hasn't been provided by the caller.
+        // Normally, we could leave this null for NumPy to allocate memory for us, but
+        // since we need a memoryview, the data pointer has to be non-null. NumPy uses
+        // malloc if NPY_NEEDS_INIT is not set (in which case it uses calloc); however,
+        // we don't have a descriptor yet (only a buffer format string), so we can't
+        // access the flags. The safest thing to do is thus to use calloc.
+        auto buf_info = info;
+        if (!buf_info.ptr)
+            // always allocate at least 1 element, same way as NumPy does it
+            buf_info.ptr = std::calloc(std::max(info.size, 1ul), info.itemsize);
```

Comment:
Do you see any way to provide a way to allocate uninitialized memory, which was what the previous implementation did? I would say that in almost all cases, the user will allocate a buffer and then fill it evaluations with some kind of numerical routine, so the initialization is just wasted.

also, should this be (size_t) 1 instead of 1ul in the std::max call?